File: src/km_g15_rgb/device.py
"""USB HID Device Communication Layer"""
import hid
from typing import Optional, List


class KM_G15_Device:
    """AUKEY KM-G15 USB HID device wrapper."""
    
    # Known VID/PID for AUKEY KM-G15 (SONiX chip)
    VENDOR_ID = 0x0C45
    PRODUCT_ID = 0x7666
    
    # Usage Pages
    USAGE_PAGE_KEYBOARD = 0x01
    USAGE_PAGE_VENDOR = 0xFF1C
    
    # Usages
    USAGE_KEYBOARD = 0x06
    USAGE_VENDOR_RGB = 0x0092
    
    # Report ID for RGB control
    REPORT_ID = 0x04
    
    def __init__(self):
        self._device: Optional[hid.device] = None
        self._vendor_device_info: Optional[dict] = None
    
    @staticmethod
    def enumerate_devices() -> List[dict]:
        """Enumerate all KM-G15 devices."""
        devices = hid.enumerate(KM_G15_Device.VENDOR_ID, KM_G15_Device.PRODUCT_ID)
        return devices
    
    @staticmethod
    def find_rgb_interface() -> Optional[dict]:
        """Find the vendor-specific RGB control interface."""
        devices = hid.enumerate(KM_G15_Device.VENDOR_ID, KM_G15_Device.PRODUCT_ID)
        for d in devices:
            if d.get('usage_page') == KM_G15_Device.USAGE_PAGE_VENDOR:
                return d
        return None
    
    def open(self, serial_number: Optional[str] = None) -> bool:
        """Open the RGB control interface."""
        rgb_interface = self.find_rgb_interface()
        if not rgb_interface:
            raise RuntimeError("RGB control interface not found. Is the keyboard connected?")

        self._vendor_device_info = rgb_interface
        self._device = hid.device()

        try:
            # Use open_path() to precisely open the RGB interface
            self._device.open_path(rgb_interface['path'])
            return True
        except Exception as e:
            self._device = None
            raise RuntimeError(f"Failed to open device: {e}")
    
    def close(self):
        """Close the device connection."""
        if self._device:
            try:
                self._device.close()
            except:
                pass
            self._device = None
    
    def write(self, data: bytes) -> int:
        """Write data to the device (Output Report)."""
        if not self._device:
            raise RuntimeError("Device not open")

        # Send raw 64-byte packet
        data_list = list(data)
        return self._device.write(data_list)
    
    def send_report(self, packet: bytes) -> bool:
        """Send a 64-byte RGB control packet."""
        if len(packet) != 64:
            raise ValueError(f"Packet must be 64 bytes, got {len(packet)}")

        result = self.write(packet)
        return result == 64

    def read(self, timeout_ms: int = 100) -> Optional[bytes]:
        """Read data from the device."""
        if not self._device:
            raise RuntimeError("Device not open")

        try:
            data = self._device.read(64, timeout_ms)
            if data:
                return bytes(data)
            return None
        except Exception:
            return None
    
    def get_manufacturer_string(self) -> Optional[str]:
        """Get the manufacturer string."""
        if not self._device:
            return None
        try:
            return self._device.get_manufacturer_string()
        except:
            return None
    
    def get_product_string(self) -> Optional[str]:
        """Get the product string."""
        if not self._device:
            return None
        try:
            return self._device.get_product_string()
        except:
            return None
    
    def __enter__(self):
        self.open()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
        return False

File: go/internal/device/device.go
// Package device provides USB HID device communication for the AUKEY KM-G15 keyboard.
package device

import (
	"fmt"
	"time"

	hid "github.com/sstallion/go-hid"
)

const (
	VendorID  = 0x0C45
	ProductID = 0x7666

	UsagePageVendor = 0xFF1C
	UsageVendorRGB  = 0x0092

	PacketSize = 64
)

// DeviceInfo holds information about a discovered HID device.
type DeviceInfo struct {
	VendorID          uint16
	ProductID         uint16
	Path              string
	ProductString     string
	ManufacturerString string
	UsagePage         uint16
	Usage             uint16
	InterfaceNumber   int
}

// Device wraps a hid.Device for KM-G15 RGB communication.
type Device struct {
	handle *hid.Device
}

// EnumerateDevices finds all KM-G15 HID interfaces.
func EnumerateDevices() ([]DeviceInfo, error) {
	var devices []DeviceInfo

	err := hid.Enumerate(VendorID, ProductID, func(info *hid.DeviceInfo) error {
		devices = append(devices, DeviceInfo{
			VendorID:          info.VendorID,
			ProductID:         info.ProductID,
			Path:              info.Path,
			ProductString:     info.ProductStr,
			ManufacturerString: info.MfrStr,
			UsagePage:         info.UsagePage,
			Usage:             info.Usage,
			InterfaceNumber:   info.InterfaceNbr,
		})
		return nil
	})
	if err != nil {
		return nil, fmt.Errorf("hid enumerate: %w", err)
	}
	return devices, nil
}

// FindRGBInterface finds the vendor-specific RGB control interface.
func FindRGBInterface() (*DeviceInfo, error) {
	devices, err := EnumerateDevices()
	if err != nil {
		return nil, err
	}

	for i := range devices {
		if devices[i].UsagePage == UsagePageVendor && devices[i].Usage == UsageVendorRGB {
			return &devices[i], nil
		}
	}
	return nil, fmt.Errorf("RGB control interface not found (usage page 0x%04X, usage 0x%04X) - is the keyboard connected?",
		UsagePageVendor, UsageVendorRGB)
}

// Open opens the RGB control interface by device path.
func Open(path string) (*Device, error) {
	handle, err := hid.OpenPath(path)
	if err != nil {
		return nil, fmt.Errorf("open device: %w", err)
	}
	return &Device{handle: handle}, nil
}

// OpenDefault finds and opens the first RGB interface.
func OpenDefault() (*Device, error) {
	info, err := FindRGBInterface()
	if err != nil {
		return nil, err
	}
	return Open(info.Path)
}

// Write sends raw data to the device.
func (d *Device) Write(data []byte) (int, error) {
	if d.handle == nil {
		return 0, fmt.Errorf("device not open")
	}
	n, err := d.handle.Write(data)
	if err != nil {
		return 0, fmt.Errorf("write: %w", err)
	}
	return n, nil
}

// SendReport sends a 64-byte RGB control packet.
func (d *Device) SendReport(packet []byte) error {
	if len(packet) != PacketSize {
		return fmt.Errorf("packet must be %d bytes, got %d", PacketSize, len(packet))
	}
	n, err := d.Write(packet)
	if err != nil {
		return err
	}
	if n != PacketSize {
		return fmt.Errorf("wrote %d bytes, expected %d", n, PacketSize)
	}
	return nil
}

// Read reads data from the device with a timeout.
func (d *Device) Read(timeoutMs int) ([]byte, error) {
	if d.handle == nil {
		return nil, fmt.Errorf("device not open")
	}
	buf := make([]byte, PacketSize)
	n, err := d.handle.ReadWithTimeout(buf, time.Duration(timeoutMs)*time.Millisecond)
	if err != nil {
		return nil, nil // Timeout or error treated as no data
	}
	if n == 0 {
		return nil, nil
	}
	return buf[:n], nil
}

// GetManufacturerString returns the manufacturer string.
func (d *Device) GetManufacturerString() string {
	if d.handle == nil {
		return ""
	}
	s, err := d.handle.GetMfrStr()
	if err != nil {
		return ""
	}
	return s
}

// GetProductString returns the product string.
func (d *Device) GetProductString() string {
	if d.handle == nil {
		return ""
	}
	s, err := d.handle.GetProductStr()
	if err != nil {
		return ""
	}
	return s
}

// Close closes the device connection.
func (d *Device) Close() error {
	if d.handle != nil {
		err := d.handle.Close()
		d.handle = nil
		return err
	}
	return nil
}
